lib/lhm/progress.rb

```diff
@@ -0,0 +1,202 @@
+require "lhm/printer"
+require "json"
+
+module Lhm
+  STATE_COPYING = "copying"
+  STATE_COPYING_DONE = "copying_done"
+  STATE_COPYING_FAILED = "copying_failed"
+  STATE_INITIAL = "initial"
+  STATE_SETUP_FAILED = "setup_failed"
+  STATE_SETUP_DONE = "setup_done"
+  STATE_SWITCHING_TABLES = "switching_tables"
+  STATE_SWITCHED_TABLES = "switched_tables"
+  STATE_SWITCHING_TABLES_FAILED = "switching_tables_failed"
+  STATE_TRIGGERS_DROPPED = "trigggers_dropeed"
+
+  class Progress
+    def initialize(origin, connection = nil, options: {})
+      @origin = origin
+      @connection = connection
+      @options = options
+      @throttler = options.include?(:throttler) ? options[:throttler].class.name : "None"
+      @table_switcher = options.include?(:atomic_switch) ? "AtomicSwitcher" : "LockedSwitcher"
+      @state = Lhm::STATE_INITIAL
+      @printer = options[:printer] || Printer::Percentage.new
+  
+      @rows_written = 0    
+      @speedometer_window = options[:speedometer_window] || 5 * 60
+      @speedometer = Speedometer.new(@speedometer_window)
+      @copy_speed = 0.0
+      @bytes_copied = 0
+      @avg_row_length = 0
+      @total_bytes = 0
+      @completion_percentage = 0
+      @start_time = Time.now
+
+      @retry_helper = SqlRetry.new(
+        @connection,
+        {
+          log_prefix: "Progress"
+        }.merge!(options.fetch(:retriable, {}))
+      )
+    end
+
+    attr_reader :state, :completion_percentage
+
+    def update_before_copy(start, limit)
+      populate_table_stats
```

Comment:
Nitpick: since this method is only ever used once, and it is somewhat cryptically named (what is table stats and what are we populating it with?), can we just move the content of the 3 lines of code of this method directly inline here?